synapse/handlers/federation_event.py

```diff
@@ -906,6 +906,11 @@ async def _process_pulled_event(
 
             if e.code == 403:
                 logger.warning("Pulled event %s failed history check.", event_id)
+            elif e.code == 429:
+                logger.warning(
+                    "Not attempting to pull event %s that we already tried to pull recently (backing off).",
+                    event_id,
+                )
```

Comment:
Perhaps we shouldn't record a failure when it's a 429 when we're backing off fetching one of this events `prev_events`. Not being able to fetch the `prev_events` for this `eventA`, doesn't mean we won't be able to fetch `eventA` as a `prev_event` for another downstream event.

Otherwise, this currently causes a cascade of failures through the whole DAG.

<details>
<summary>Example cascade from <code>logs/server-0/client_reader.log</code> running Sytest <code>tests/48admin.pl</code> "Can backfill purged history"</summary>

```
2022-10-05 08:38:33,218 - synapse.http.matrixfederationclient - 308 - INFO - GET-4 - {GET-O-1} [localhost:8879] Completed request: 200 OK in 0.02 secs, got 7821 bytes - GET matrix://localhost:8879/_matrix/federation/v1/backfill/%21NUgUcBPsJkCbellfTQ%3Alocalhost%3A8839?v=%24Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk&limit=100
2022-10-05 08:38:33,230 - synapse.handlers.federation_event - 826 - INFO - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 - Processing pulled event <FrozenEventV3 event_id=$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78, type=m.room.member, state_key=@anon-20221005_083744-7:localhost:8879, outlier=False>
2022-10-05 08:38:33,234 - synapse.handlers.federation_event - 859 - INFO - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 - De-outliering event $Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78
2022-10-05 08:38:33,243 - synapse.handlers.federation_event - 973 - INFO - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 - _compute_event_context_with_maybe_missing_prevs(event=$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78) prevs_to_ignore=[]
2022-10-05 08:38:33,243 - synapse.handlers.federation_event - 990 - INFO - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 - Event $Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 is missing prev_events ['$XCQvvErHXkLvYbzgEgKI-MTDN6HsFbFzbAcD5yX9U44']: calculating state for a backwards extremity
2022-10-05 08:38:33,244 - synapse.handlers.federation_event - 1016 - INFO - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 - Requesting state after missing prev_event $XCQvvErHXkLvYbzgEgKI-MTDN6HsFbFzbAcD5yX9U44
2022-10-05 08:38:33,257 - synapse.http.matrixfederationclient - 634 - INFO - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78-$XCQvvErHXkLvYbzgEgKI-MTDN6HsFbFzbAcD5yX9U44 - {GET-O-2} [localhost:8879] Got response headers: 404 Not Found
2022-10-05 08:38:33,257 - synapse.http.matrixfederationclient - 715 - WARNING - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78-$XCQvvErHXkLvYbzgEgKI-MTDN6HsFbFzbAcD5yX9U44 - {GET-O-2} [localhost:8879] Request failed: GET matrix://localhost:8879/_matrix/federation/v1/state_ids/%21NUgUcBPsJkCbellfTQ%3Alocalhost%3A8839?event_id=%24XCQvvErHXkLvYbzgEgKI-MTDN6HsFbFzbAcD5yX9U44: HttpResponseException('404: Not Found')
2022-10-05 08:38:33,258 - synapse.handlers.federation_event - 1042 - WARNING - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 - Error attempting to resolve state at missing prev_events
Traceback (most recent call last):
  File "/synapse/synapse/handlers/federation_event.py", line 1024, in _compute_event_context_with_maybe_missing_prevs
    dest, room_id, p
  File "/synapse/synapse/logging/opentracing.py", line 889, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/synapse/synapse/logging/opentracing.py", line 889, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/synapse/synapse/handlers/federation_event.py", line 1088, in _get_state_ids_after_missing_prev_event
    destination, room_id, event_id=event_id
  File "/synapse/synapse/logging/opentracing.py", line 889, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/synapse/synapse/logging/opentracing.py", line 889, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/synapse/synapse/federation/federation_client.py", line 479, in get_room_state_ids
    destination, room_id, event_id=event_id
  File "/synapse/synapse/federation/transport/client.py", line 81, in get_room_state_ids
    try_trailing_slash_on_400=True,
  File "/synapse/synapse/http/matrixfederationclient.py", line 1084, in get_json
    timeout=timeout,
  File "/synapse/synapse/http/matrixfederationclient.py", line 406, in _send_request_with_optional_trailing_slash
    response = await self._send_request(request, **send_request_args)
  File "/synapse/synapse/http/matrixfederationclient.py", line 668, in _send_request
    raise exc
synapse.api.errors.HttpResponseException: 404: Not Found
2022-10-05 08:38:33,259 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78, cause=ERROR 403: We can't get valid state history.
2022-10-05 08:38:33,262 - synapse.handlers.federation_event - 908 - WARNING - GET-4-$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 - Pulled event $Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 failed history check.
2022-10-05 08:38:33,262 - synapse.handlers.federation_event - 826 - INFO - GET-4-$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg - Processing pulled event <FrozenEventV3 event_id=$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg, type=m.room.name, state_key=, outlier=False>
2022-10-05 08:38:33,265 - synapse.handlers.federation_event - 859 - INFO - GET-4-$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg - De-outliering event $tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg
2022-10-05 08:38:33,269 - synapse.handlers.federation_event - 973 - INFO - GET-4-$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg - _compute_event_context_with_maybe_missing_prevs(event=$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg) prevs_to_ignore=['$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78']
2022-10-05 08:38:33,269 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg, cause=ERROR 429: While computing context for event=$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg, not attempting to pull missing prev_event=$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 because we already tried to pull recently (backing off).
2022-10-05 08:38:33,271 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg - Not attempting to pull event=$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg because of affected=$Dwi6IsKNmb9Zip89IuoDlZ6Wule6ke7YtavutZQ2o78 that we already tried to pull recently (backing off).
2022-10-05 08:38:33,271 - synapse.handlers.federation_event - 826 - INFO - GET-4-$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8 - Processing pulled event <FrozenEventV3 event_id=$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8, type=m.room.message, state_key=None, outlier=False>
2022-10-05 08:38:33,277 - synapse.handlers.federation_event - 973 - INFO - GET-4-$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8 - _compute_event_context_with_maybe_missing_prevs(event=$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8) prevs_to_ignore=['$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg']
2022-10-05 08:38:33,277 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8 - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8, cause=ERROR 429: While computing context for event=$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8, not attempting to pull missing prev_event=$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg because we already tried to pull recently (backing off).
2022-10-05 08:38:33,279 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8 - Not attempting to pull event=$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8 because of affected=$tPGVn9ZpXEPcWznLECrIL0HDPxqlBz1RFB8mZ3YKmGg that we already tried to pull recently (backing off).
2022-10-05 08:38:33,279 - synapse.handlers.federation_event - 826 - INFO - GET-4-$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U - Processing pulled event <FrozenEventV3 event_id=$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U, type=m.room.message, state_key=None, outlier=False>
2022-10-05 08:38:33,285 - synapse.handlers.federation_event - 973 - INFO - GET-4-$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U - _compute_event_context_with_maybe_missing_prevs(event=$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U) prevs_to_ignore=['$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8']
2022-10-05 08:38:33,285 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U, cause=ERROR 429: While computing context for event=$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U, not attempting to pull missing prev_event=$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8 because we already tried to pull recently (backing off).
2022-10-05 08:38:33,287 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U - Not attempting to pull event=$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U because of affected=$XghVHINVOPDZ-QyfzjGZECDMlKBFjHvllhIR7kqemT8 that we already tried to pull recently (backing off).
2022-10-05 08:38:33,287 - synapse.handlers.federation_event - 826 - INFO - GET-4-$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk - Processing pulled event <FrozenEventV3 event_id=$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk, type=m.room.message, state_key=None, outlier=False>
2022-10-05 08:38:33,292 - synapse.handlers.federation_event - 973 - INFO - GET-4-$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk - _compute_event_context_with_maybe_missing_prevs(event=$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk) prevs_to_ignore=['$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U']
2022-10-05 08:38:33,292 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk, cause=ERROR 429: While computing context for event=$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk, not attempting to pull missing prev_event=$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U because we already tried to pull recently (backing off).
2022-10-05 08:38:33,293 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk - Not attempting to pull event=$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk because of affected=$_uS1OEZB1hMQlDnKQOWLla1tNDoLC4OjlZtpV4ke02U that we already tried to pull recently (backing off).
2022-10-05 08:38:33,294 - synapse.handlers.federation_event - 826 - INFO - GET-4-$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U - Processing pulled event <FrozenEventV3 event_id=$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U, type=m.room.message, state_key=None, outlier=False>
2022-10-05 08:38:33,298 - synapse.handlers.federation_event - 973 - INFO - GET-4-$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U - _compute_event_context_with_maybe_missing_prevs(event=$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U) prevs_to_ignore=['$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk']
2022-10-05 08:38:33,298 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U, cause=ERROR 429: While computing context for event=$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U, not attempting to pull missing prev_event=$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk because we already tried to pull recently (backing off).
2022-10-05 08:38:33,299 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U - Not attempting to pull event=$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U because of affected=$0xaQzHqGsK0unAXRX6b9U6nkNIKTTq_NXrYi7CCX6Vk that we already tried to pull recently (backing off).
2022-10-05 08:38:33,299 - synapse.handlers.federation_event - 826 - INFO - GET-4-$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw - Processing pulled event <FrozenEventV3 event_id=$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw, type=m.room.message, state_key=None, outlier=False>
2022-10-05 08:38:33,304 - synapse.handlers.federation_event - 973 - INFO - GET-4-$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw - _compute_event_context_with_maybe_missing_prevs(event=$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw) prevs_to_ignore=['$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U']
2022-10-05 08:38:33,304 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw, cause=ERROR 429: While computing context for event=$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw, not attempting to pull missing prev_event=$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U because we already tried to pull recently (backing off).
2022-10-05 08:38:33,305 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw - Not attempting to pull event=$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw because of affected=$MCjcVp__YCCorEi3FjoMa-k9WMIu71NdXqoc8P2862U that we already tried to pull recently (backing off).
2022-10-05 08:38:33,305 - synapse.handlers.federation_event - 826 - INFO - GET-4-$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc - Processing pulled event <FrozenEventV3 event_id=$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc, type=m.room.message, state_key=None, outlier=False>
2022-10-05 08:38:33,310 - synapse.handlers.federation_event - 973 - INFO - GET-4-$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc - _compute_event_context_with_maybe_missing_prevs(event=$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc) prevs_to_ignore=['$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw']
2022-10-05 08:38:33,310 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc, cause=ERROR 429: While computing context for event=$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc, not attempting to pull missing prev_event=$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw because we already tried to pull recently (backing off).
2022-10-05 08:38:33,312 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc - Not attempting to pull event=$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc because of affected=$xtZzN_ER45gozEm62m1ChGyNoaiWaNDNVY_r9iMBiqw that we already tried to pull recently (backing off).
2022-10-05 08:38:33,312 - synapse.handlers.federation_event - 826 - INFO - GET-4-$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU - Processing pulled event <FrozenEventV3 event_id=$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU, type=m.room.message, state_key=None, outlier=False>
2022-10-05 08:38:33,317 - synapse.handlers.federation_event - 973 - INFO - GET-4-$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU - _compute_event_context_with_maybe_missing_prevs(event=$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU) prevs_to_ignore=['$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc']
2022-10-05 08:38:33,318 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU, cause=ERROR 429: While computing context for event=$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU, not attempting to pull missing prev_event=$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc because we already tried to pull recently (backing off).
2022-10-05 08:38:33,319 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU - Not attempting to pull event=$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU because of affected=$HU9oV1TTzArEoqIUSl52TR58msts3A0xFxW1p5gBXVc that we already tried to pull recently (backing off).
2022-10-05 08:38:33,320 - synapse.handlers.federation_event - 826 - INFO - GET-4-$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM - Processing pulled event <FrozenEventV3 event_id=$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM, type=m.room.message, state_key=None, outlier=False>
2022-10-05 08:38:33,325 - synapse.handlers.federation_event - 973 - INFO - GET-4-$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM - _compute_event_context_with_maybe_missing_prevs(event=$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM) prevs_to_ignore=['$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU']
2022-10-05 08:38:33,325 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM, cause=ERROR 429: While computing context for event=$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM, not attempting to pull missing prev_event=$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU because we already tried to pull recently (backing off).
2022-10-05 08:38:33,326 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM - Not attempting to pull event=$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM because of affected=$guHvD5lD3sZEQZLcu48N1m2qEjJ5VFVxVxH3U66RSJU that we already tried to pull recently (backing off).
2022-10-05 08:38:33,327 - synapse.handlers.federation_event - 826 - INFO - GET-4-$Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk - Processing pulled event <FrozenEventV3 event_id=$Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk, type=m.room.message, state_key=None, outlier=False>
2022-10-05 08:38:33,332 - synapse.handlers.federation_event - 973 - INFO - GET-4-$Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk - _compute_event_context_with_maybe_missing_prevs(event=$Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk) prevs_to_ignore=['$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM']
2022-10-05 08:38:33,332 - synapse.storage.databases.main.event_federation - 1508 - INFO - GET-4-$Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk - record_event_failed_pull_attempt room_id=!NUgUcBPsJkCbellfTQ:localhost:8839, event_id=$Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk, cause=ERROR 429: While computing context for event=$Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk, not attempting to pull missing prev_event=$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM because we already tried to pull recently (backing off).
2022-10-05 08:38:33,333 - synapse.handlers.federation_event - 913 - WARNING - GET-4-$Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk - Not attempting to pull event=$Gos2iAJL9b7Nh7z2Uk9QUpsWmelgfIk-HFIU6pOFiXk because of affected=$v3sNtiMJyAqNg2c3eCyic7RklAhVQkyqYi_eHQvP1RM that we already tried to pull recently (backing off).
```

</details>

Reply:
> Not being able to fetch the `prev_events` for this `eventA`, doesn't mean we won't be able to fetch `eventA` as a `prev_event` for another downstream event (is this actually true? is there anything special we can do with state/auth events).

Reviewer: Please double-check this.

Basically, if we can't pull a `prev_event` for an event, we can't de-outlier it (see [`outlier` definition](https://github.com/matrix-org/synapse/blob/release-v1.68/docs/development/room-dag-concepts.md#outliers)).

But we can still use an `outlier` in the state/auth chain for another event. So we shouldn't stop a downstream event from trying to pull it.


For example with these two scenarios, if we process the bolded event and are unable to pull the event before it, can we still use the bolded event when processing the event after?

 1. `invite` <- **`join`** <- `message`
 1. `message1` <- **`message2`** <- `message3`